core/arch/arm/sta/gprof.c

```diff
@@ -0,0 +1,275 @@
+
+#include <arm.h>
+#include <kernel/misc.h>
+#include <kernel/static_ta.h>
+#include <kernel/user_ta.h>
+#include <kernel/thread.h>
+#include <mm/core_memprot.h>
+#include <mm/tee_mmu.h>
+#include <optee_msg_supplicant.h>
+#include <pta_gprof.h>
+#include <string.h>
+
+static TEE_Result gprof_send_rpc(TEE_UUID *uuid, void *buf, size_t len,
+				 uint32_t *id)
+{
+	struct optee_msg_param params[3];
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint64_t c = 0;
+	paddr_t pa;
+	char *va;
+
+	thread_rpc_alloc_payload(sizeof(*uuid) + len, &pa, &c);
+	if (!pa)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	va = phys_to_virt(pa, MEM_AREA_NSEC_SHM);
+	if (!va)
+		goto exit;
+
+	memcpy(va, uuid, sizeof(*uuid));
+	memcpy(va + sizeof(*uuid), buf, len);
+
+	memset(params, 0, sizeof(params));
+	params[0].attr = OPTEE_MSG_ATTR_TYPE_VALUE_INOUT;
+	params[0].u.value.a = *id;
+
+	params[1].attr = OPTEE_MSG_ATTR_TYPE_TMEM_INPUT;
+	params[1].u.tmem.buf_ptr = pa;
+	params[1].u.tmem.size = sizeof(*uuid);
+	params[1].u.tmem.shm_ref = c;
+
+	params[2].attr = OPTEE_MSG_ATTR_TYPE_TMEM_INPUT;
+	params[2].u.tmem.buf_ptr = pa + sizeof(*uuid);
+	params[2].u.tmem.size = len;
+	params[2].u.tmem.shm_ref = c;
+
+	res = thread_rpc_cmd(OPTEE_MSG_RPC_CMD_GPROF, 3, params);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	*id = (uint32_t)params[0].u.value.a;
+exit:
+	thread_rpc_free_payload(c);
+	return res;
+}
+
+/* Return the TA session that invoked us through the TA-TA API */
+static TEE_Result get_calling_session(struct tee_ta_session **s)
+{
+	TEE_Result res;
+
+	res = tee_ta_get_current_session(s);
+	if (res != TEE_SUCCESS)
+		return res;
+	*s = tee_ta_get_calling_session();
+	return TEE_SUCCESS;
+}
+
+static TEE_Result gprof_send(uint32_t param_types,
+			     TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INOUT,
+					  TEE_PARAM_TYPE_MEMREF_INPUT,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE);
+	struct tee_ta_session *s;
+	TEE_Result res;
+
+	if (exp_pt != param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = get_calling_session(&s);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	return gprof_send_rpc(&s->ctx->uuid, params[1].memref.buffer,
+			      params[1].memref.size, &params[0].value.a);
+}
+
+static TEE_Result gprof_start_pc_sampling(uint32_t param_types,
+					  TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					  TEE_PARAM_TYPE_VALUE_INPUT,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE);
+	struct tee_ta_session *s;
+	struct sample_buf *sbuf;
+	TEE_Result res;
+	uint32_t len;
+	uint32_t offset;
+	uint32_t scale;
+
+	if (exp_pt != param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = get_calling_session(&s);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	len = params[0].value.a;
+	offset = params[0].value.b;
+	scale = params[1].value.a;
+
+	sbuf = calloc(1, sizeof(*sbuf));
+	if (!sbuf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	sbuf->samples = calloc(1, len);
```

Comment:
Wouldn't it be nice to keep this buffer in user space too? Especially since user space decides the size.